### ISSUE TITLE
Adjacent rules with `:hover::after` don't get merged

### ISSUE THREAD

**Issue by ngyikp** [jakubpawlowicz/clean-css#886]
Input:
```css
.a:hover:after {color: #000;}
.b:after {color: #000;}

.c:hover::after {color: #111;}
.d:hover::after {color: #111;}
.e::after {color: #111;}
```

Output from clean-css 3.4.24:
```
$ ./node_modules/.bin/cleancss in.css

.a:hover:after,.b:after{color:#000}.c:hover::after,.d:hover::after,.e::after{color:#111}
```

Output from clean-css 4.0.5:
```
$ ./node_modules/.bin/cleancss in.css -O2

.a:hover:after,.b:after{color:#000}.c:hover::after{color:#111}.d:hover::after{color:#111}.e::after{color:#111}
```

**Comment by jakubpawlowicz** [comment]
That's #829 which restricted it. Are we a bit too strict here?

**Comment by jakubpawlowicz** [comment]
So I think this is actually correct now since `:hover::after` doesn't to work correctly on IE10, our default compatibility mode. However it works fine on IE11 so it can be merged with `::after` - the question is how about other browsers. See https://jsfiddle.net/4wrbL5o2/

**Comment by ngyikp** [comment]
On my testbed of old Chrome versions, it's indeed broken on Chrome 12 (mousing over works, but it doesn't revert the style on mouse out) but Chrome 17 works fine. Unfortunately I don't have versions 13-15 to test.

At https://developer.mozilla.org/en/docs/Web/CSS/:hover#Browser_compatibility , it says Firefox 28 is needed, but it works fine starting with Firefox 6 or even older if you set a proper doctype (https://developer.mozilla.org/en-US/Firefox/Releases/6#Other_changes)

**Comment by jakubpawlowicz** [comment]
What matters here is whether `div`s 1-2 and 3-4 works in the same way. If they do then `::after` and `:hover::after` can be merged.

**Comment by ngyikp** [comment]
Yeah, they work identically on my tests

**Comment by jakubpawlowicz** [comment]
👍 I've just realized it works identically on IE10 too - I mean `:hover::after` doesn't work at all either if left alone or merged with `::after` - so we can actually merge those selectors.

**Comment by jakubpawlowicz** [comment]
@ngyikp could you please give https://jsfiddle.net/xr82fnre/2/ a try in older Chrome / Firefox? It should give you red "12" in output.

What I'm doing there is trying various combinations of pseudo classes and elements to figure out if they can be mixed together. So far I found out that mixing any of `:after`, `:before`, `:first-letter`, `:first-line`, and `:lang` is not allowed.

**Comment by ngyikp** [comment]
Breaks on Firefox 5 and I don't have even older Chrome versions

![](https://cloud.githubusercontent.com/assets/6135313/22810125/d24e960e-ef71-11e6-8ba6-af1b32639b6c.png)

**Comment by jakubpawlowicz** [comment]
Thanks a ton for checking! FF5 being broken would mean one of pseudo classes is not supported.

We should restore this feature but I'd also like to have a flag to disable this optimization.

**Comment by jakubpawlowicz** [comment]
It's ready ^ on master.

**Comment by jakubpawlowicz** [comment]
It's been released in 4.1.
